src/libraries/Microsoft.PowerFx.Interpreter/RecalcEngine.cs

```diff
@@ -207,48 +209,27 @@ internal DefineFunctionsResult DefineFunctions(string script, bool numberIsFloat
         }
 
         [Obsolete("preview")]
-        internal void DefineType(string script, ParserOptions parserOptions)
+        internal IEnumerable<TexlError> DefineType(string script, ParserOptions parserOptions)
         {
             var parsedNamedFormulasAndUDFs = UserDefinitions.Parse(script, parserOptions);
             var definedTypes = parsedNamedFormulasAndUDFs.DefinedTypes.ToList();
 
+            var errors = new List<TexlError>();
+
             foreach (var defType in definedTypes)
             {
                 var name = defType.Ident.Name.Value;
-                var res = DTypeFromTexlNode(defType.Type.TypeRoot) ?? throw new Exception("Failed defining type");
-                _definedTypeSymbolTable.RegisterType(name, new KnownRecordType(res));
-            }
-        }
-
-        internal DType DTypeFromTexlNode(TexlNode node)
-        {
-            if (node is RecordNode recordNode)
-            {
-                return DTypeFromRecordNode(recordNode);
-            }
-            else if (node is FirstNameNode nameNode)
-            {
-                return GetTypeFromName(nameNode.Ident.Name.Value);
-            }
-
-            return null;
-        }
-
-        internal DType DTypeFromRecordNode(RecordNode recordNode)
-        {
-            var list = new List<TypedName>();
-            foreach (var (node, ident) in recordNode.ChildNodes.Zip(recordNode.Ids, (a, b) => (a, b)))
-            {
-                var ty = DTypeFromTexlNode(node);
-                if (ty == null)
+                var res = defType.Type.TypeRoot.Accept(new DTypeVisitor(), _definedTypeSymbolTable);

```

Comment:
Nit:
in general, for visitors, we have a static `Run` method that constructs the instance, and calls accept, and a private constructor. This allows more control over how/where it's invoked